File: core/uds.py
import struct
import time
from enum import IntEnum
from typing import Optional
from core.j2534 import J2534, Protocol, FilterType
from core.protocols import NetworkConfig


class UDSService(IntEnum):
    DIAGNOSTIC_SESSION_CONTROL = 0x10
    ECU_RESET = 0x11
    CLEAR_DTC = 0x14
    READ_DTC_INFO = 0x19
    READ_DATA_BY_ID = 0x22
    READ_MEMORY_BY_ADDRESS = 0x23
    SECURITY_ACCESS = 0x27
    COMMUNICATION_CONTROL = 0x28
    WRITE_DATA_BY_ID = 0x2E
    IO_CONTROL = 0x2F
    ROUTINE_CONTROL = 0x31
    REQUEST_DOWNLOAD = 0x34
    TRANSFER_DATA = 0x36
    REQUEST_TRANSFER_EXIT = 0x37
    TESTER_PRESENT = 0x3E


class UDSSession(IntEnum):
    DEFAULT = 0x01
    PROGRAMMING = 0x02
    EXTENDED = 0x03
    FORD_DIAG = 0x85


class DTCSubFunction(IntEnum):
    REPORT_NUMBER_BY_STATUS = 0x01
    REPORT_BY_STATUS = 0x02
    REPORT_SNAPSHOT_ID = 0x03
    REPORT_SNAPSHOT_BY_DTC = 0x04
    REPORT_STORED_DATA = 0x06
    REPORT_PENDING = 0x07
    REPORT_CONFIRMED = 0x0A
    REPORT_SUPPORTED_DTC = 0x0F


class NRC(IntEnum):
    GENERAL_REJECT = 0x10
    SERVICE_NOT_SUPPORTED = 0x11
    SUBFUNCTION_NOT_SUPPORTED = 0x12
    INCORRECT_LENGTH = 0x13
    RESPONSE_TOO_LONG = 0x14
    BUSY_REPEAT = 0x21
    CONDITIONS_NOT_CORRECT = 0x22
    REQUEST_SEQUENCE_ERROR = 0x24
    REQUEST_OUT_OF_RANGE = 0x31
    SECURITY_ACCESS_DENIED = 0x33
    INVALID_KEY = 0x35
    EXCEEDED_ATTEMPTS = 0x36
    TIME_DELAY_NOT_EXPIRED = 0x37
    UPLOAD_DOWNLOAD_NOT_ACCEPTED = 0x70
    TRANSFER_SUSPENDED = 0x71
    GENERAL_PROGRAMMING_FAILURE = 0x72
    RESPONSE_PENDING = 0x78
    SUBFUNCTION_NOT_SUPPORTED_IN_SESSION = 0x7E
    SERVICE_NOT_SUPPORTED_IN_SESSION = 0x7F


class UDSException(Exception):
    def __init__(self, service: int, nrc: int):
        self.service = service
        self.nrc = nrc
        nrc_name = NRC(nrc).name if nrc in NRC._value2member_map_ else f"0x{nrc:02X}"
        svc_name = UDSService(service).name if service in UDSService._value2member_map_ else f"0x{service:02X}"
        super().__init__(f"Negative response to {svc_name}: {nrc_name}")


# Ford / J2534 ISO-TP constants observed in an external Ford-diagnostic
# reverse-engineering reference. The 3001ms upper bound matches the
# MAX_RESPONSE_TIME setting Ford's tool uses — long enough for slow ECU
# operations (BCM config writes, TCM programming sessions) without
# false-timing-out a healthy bus.
DEFAULT_REQUEST_TIMEOUT_MS = 3001
ISO_TP_BLOCK_SIZE_FRAMES   = 200    # max consecutive frames between flow-control


class UDSClient:
    def __init__(self, j2534: J2534, network: NetworkConfig, tx_id: int, rx_id: int):
        self.j2534 = j2534
        self.network = network
        self.tx_id = tx_id
        self.rx_id = rx_id
        self.channel_id: Optional[int] = None
        self.filter_id: Optional[int] = None
        self.tester_present_id: Optional[int] = None
        self.timeout = DEFAULT_REQUEST_TIMEOUT_MS

    def connect(self):
        self.channel_id = self.j2534.connect(
            self.network.protocol, self.network.flags, self.network.baudrate
        )
        if self.network.can_id_bits == 11:
            mask = struct.pack(">I", 0x7FF)
            pattern = struct.pack(">I", self.rx_id)
            fc = struct.pack(">I", self.tx_id)
        else:
            mask = struct.pack(">I", 0x1FFFFFFF)
            pattern = struct.pack(">I", self.rx_id)
            fc = struct.pack(">I", self.tx_id)
        self.filter_id = self.j2534.start_msg_filter(
            self.channel_id, FilterType.FLOW_CONTROL, mask, pattern, fc
        )

    def disconnect(self):
        if self.tester_present_id is not None:
            try:
                self.j2534.stop_periodic_msg(self.channel_id, self.tester_present_id)
            except Exception:
                pass
            self.tester_present_id = None
        if self.filter_id is not None:
            try:
                self.j2534.stop_msg_filter(self.channel_id, self.filter_id)
            except Exception:
                pass
            self.filter_id = None
        if self.channel_id is not None:
            try:
                self.j2534.disconnect(self.channel_id)
            except Exception:
                pass
            self.channel_id = None

    def send_raw(self, data: bytes):
        id_bytes = struct.pack(">I", self.tx_id)
        self.j2534.write_msg(self.channel_id, id_bytes + data, self.network.protocol, timeout=self.timeout)

    def receive_raw(self, timeout: Optional[int] = None) -> Optional[bytes]:
        t = timeout or self.timeout
        deadline = time.time() + t / 1000.0
        while time.time() < deadline:
            remaining = max(50, int((deadline - time.time()) * 1000))
            msgs = self.j2534.read_msgs(self.channel_id, count=1, timeout=remaining)
            for msg in msgs:
                if len(msg) < 4:
                    continue
                msg_id = struct.unpack(">I", msg[:4])[0]
                if msg_id == self.rx_id:
                    return msg[4:]
        return None

    def request(self, service: int, data: bytes = b"", timeout: Optional[int] = None) -> bytes:
        self.send_raw(bytes([service]) + data)
        while True:
            resp = self.receive_raw(timeout)
            if resp is None:
                raise TimeoutError(f"No response to service 0x{service:02X}")
            if len(resp) == 0:
                continue
            if resp[0] == 0x7F and len(resp) >= 3:
                if resp[2] == NRC.RESPONSE_PENDING:
                    continue
                raise UDSException(resp[1], resp[2])
            if resp[0] == service + 0x40:
                return resp[1:]
        raise TimeoutError(f"No valid response to service 0x{service:02X}")

    def diagnostic_session(self, session: int = UDSSession.EXTENDED):
        self.request(UDSService.DIAGNOSTIC_SESSION_CONTROL, bytes([session]))

    def tester_present(self):
        self.request(UDSService.TESTER_PRESENT, b"\x00")

    def start_tester_present(self, interval_ms: int = 2000):
        id_bytes = struct.pack(">I", self.tx_id)
        tp_data = id_bytes + bytes([UDSService.TESTER_PRESENT, 0x00])
        self.tester_present_id = self.j2534.start_periodic_msg(
            self.channel_id, tp_data, self.network.protocol, interval_ms
        )

    def ecu_reset(self, reset_type: int = 0x01):
        self.request(UDSService.ECU_RESET, bytes([reset_type]))

    def read_data_by_id(self, did: int) -> bytes:
        resp = self.request(UDSService.READ_DATA_BY_ID, struct.pack(">H", did))
        return resp[2:]

    def write_data_by_id(self, did: int, data: bytes):
        self.request(UDSService.WRITE_DATA_BY_ID, struct.pack(">H", did) + data)

    def security_access_seed(self, level: int = 0x01) -> bytes:
        resp = self.request(UDSService.SECURITY_ACCESS, bytes([level]))
        return resp[1:]

    def security_access_key(self, level: int = 0x02, key: bytes = b"") -> bytes:
        resp = self.request(UDSService.SECURITY_ACCESS, bytes([level]) + key)
        return resp[1:]

    def clear_dtc(self, group: int = 0xFFFFFF):
        self.request(UDSService.CLEAR_DTC, struct.pack(">I", group)[1:])

    def read_dtc(self, sub_function: int = DTCSubFunction.REPORT_BY_STATUS,
                 status_mask: int = 0xFF) -> bytes:
        return self.request(UDSService.READ_DTC_INFO, bytes([sub_function, status_mask]))

    def routine_control(self, routine_id: int, control_type: int = 0x01,
                        data: bytes = b"") -> bytes:
        return self.request(
            UDSService.ROUTINE_CONTROL,
            bytes([control_type]) + struct.pack(">H", routine_id) + data,
        )

    def io_control(self, did: int, control_param: int, data: bytes = b"") -> bytes:
        return self.request(
            UDSService.IO_CONTROL,
            struct.pack(">H", did) + bytes([control_param]) + data,
        )

    def __enter__(self):
        self.connect()
        return self

    def __exit__(self, *args):
        self.disconnect()

File: zig/src/uds.zig
const std = @import("std");
const j2534 = @import("j2534.zig");

pub const UDSService = enum(u8) {
    DIAGNOSTIC_SESSION_CONTROL = 0x10,
    ECU_RESET = 0x11,
    CLEAR_DTC = 0x14,
    READ_DTC_INFO = 0x19,
    READ_DATA_BY_ID = 0x22,
    READ_MEMORY_BY_ADDRESS = 0x23,
    SECURITY_ACCESS = 0x27,
    COMMUNICATION_CONTROL = 0x28,
    WRITE_DATA_BY_ID = 0x2E,
    IO_CONTROL = 0x2F,
    ROUTINE_CONTROL = 0x31,
    REQUEST_DOWNLOAD = 0x34,
    TRANSFER_DATA = 0x36,
    REQUEST_TRANSFER_EXIT = 0x37,
    TESTER_PRESENT = 0x3E,
    _,
};

pub const UDSSession = enum(u8) {
    DEFAULT = 0x01,
    PROGRAMMING = 0x02,
    EXTENDED = 0x03,
    FORD_DIAG = 0x85,
    _,
};

pub const DTCSubFunction = enum(u8) {
    REPORT_NUMBER_BY_STATUS = 0x01,
    REPORT_BY_STATUS = 0x02,
    REPORT_SNAPSHOT_ID = 0x03,
    REPORT_SNAPSHOT_BY_DTC = 0x04,
    REPORT_STORED_DATA = 0x06,
    REPORT_PENDING = 0x07,
    REPORT_CONFIRMED = 0x0A,
    REPORT_SUPPORTED_DTC = 0x0F,
    _,
};

pub const NRC = enum(u8) {
    GENERAL_REJECT = 0x10,
    SERVICE_NOT_SUPPORTED = 0x11,
    SUBFUNCTION_NOT_SUPPORTED = 0x12,
    INCORRECT_LENGTH = 0x13,
    RESPONSE_TOO_LONG = 0x14,
    BUSY_REPEAT = 0x21,
    CONDITIONS_NOT_CORRECT = 0x22,
    REQUEST_SEQUENCE_ERROR = 0x24,
    REQUEST_OUT_OF_RANGE = 0x31,
    SECURITY_ACCESS_DENIED = 0x33,
    INVALID_KEY = 0x35,
    EXCEEDED_ATTEMPTS = 0x36,
    TIME_DELAY_NOT_EXPIRED = 0x37,
    UPLOAD_DOWNLOAD_NOT_ACCEPTED = 0x70,
    TRANSFER_SUSPENDED = 0x71,
    GENERAL_PROGRAMMING_FAILURE = 0x72,
    RESPONSE_PENDING = 0x78,
    SUBFUNCTION_NOT_SUPPORTED_IN_SESSION = 0x7E,
    SERVICE_NOT_SUPPORTED_IN_SESSION = 0x7F,
    _,
};

pub const UDSClient = struct {
    j2534_inst: *j2534.J2534,
    channel_id: u32,
    tx_id: u32,
    rx_id: u32,
    protocol: j2534.Protocol,
    filter_id: ?u32 = null,

    pub fn init(j2534_inst: *j2534.J2534, channel_id: u32, tx_id: u32, rx_id: u32, protocol: j2534.Protocol) UDSClient {
        return .{
            .j2534_inst = j2534_inst,
            .channel_id = channel_id,
            .tx_id = tx_id,
            .rx_id = rx_id,
            .protocol = protocol,
        };
    }

    pub fn connect(self: *UDSClient) !void {
        if (self.protocol != .ISO15765) return error.UnsupportedProtocol;
        
        var mask_msg: j2534.PASSTHRU_MSG = undefined;
        var pattern_msg: j2534.PASSTHRU_MSG = undefined;
        var flow_msg: j2534.PASSTHRU_MSG = undefined;

        @memset(std.mem.asBytes(&mask_msg), 0);
        @memset(std.mem.asBytes(&pattern_msg), 0);
        @memset(std.mem.asBytes(&flow_msg), 0);

        mask_msg.ProtocolID = @intFromEnum(self.protocol);
        mask_msg.DataSize = 4;
        mask_msg.Data[0] = 0xFF;
        mask_msg.Data[1] = 0xFF;
        mask_msg.Data[2] = 0xFF;
        mask_msg.Data[3] = 0xFF;

        pattern_msg.ProtocolID = @intFromEnum(self.protocol);
        pattern_msg.DataSize = 4;
        pattern_msg.Data[0] = @intCast((self.rx_id >> 24) & 0xFF);
        pattern_msg.Data[1] = @intCast((self.rx_id >> 16) & 0xFF);
        pattern_msg.Data[2] = @intCast((self.rx_id >> 8) & 0xFF);
        pattern_msg.Data[3] = @intCast(self.rx_id & 0xFF);

        flow_msg.ProtocolID = @intFromEnum(self.protocol);
        flow_msg.DataSize = 4;
        flow_msg.Data[0] = @intCast((self.tx_id >> 24) & 0xFF);
        flow_msg.Data[1] = @intCast((self.tx_id >> 16) & 0xFF);
        flow_msg.Data[2] = @intCast((self.tx_id >> 8) & 0xFF);
        flow_msg.Data[3] = @intCast(self.tx_id & 0xFF);

        self.filter_id = try self.j2534_inst.startMsgFilter(self.channel_id, .FLOW_CONTROL, &mask_msg, &pattern_msg, &flow_msg);
    }

    pub fn disconnect(self: *UDSClient) !void {
        if (self.filter_id) |fid| {
            try self.j2534_inst.stopMsgFilter(self.channel_id, fid);
            self.filter_id = null;
        }
    }

    pub fn request(self: *UDSClient, req_data: []const u8, resp_buffer: []u8) !usize {
        var tx_msg: j2534.PASSTHRU_MSG = undefined;
        @memset(std.mem.asBytes(&tx_msg), 0);
        
        tx_msg.ProtocolID = @intFromEnum(self.protocol);
        tx_msg.DataSize = @as(u32, @intCast(4 + req_data.len));
        tx_msg.Data[0] = @intCast((self.tx_id >> 24) & 0xFF);
        tx_msg.Data[1] = @intCast((self.tx_id >> 16) & 0xFF);
        tx_msg.Data[2] = @intCast((self.tx_id >> 8) & 0xFF);
        tx_msg.Data[3] = @intCast(self.tx_id & 0xFF);
        
        @memcpy(tx_msg.Data[4..4 + req_data.len], req_data);

        var num_msgs: u32 = 1;
        try self.j2534_inst.writeMsgs(self.channel_id, @ptrCast(&tx_msg), &num_msgs, 1000);

        var rx_msg: j2534.PASSTHRU_MSG = undefined;
        while (true) {
            num_msgs = 1;
            @memset(std.mem.asBytes(&rx_msg), 0);
            
            self.j2534_inst.readMsgs(self.channel_id, @ptrCast(&rx_msg), &num_msgs, 1000) catch |err| switch (err) {
                error.Timeout => return error.Timeout,
                else => return err,
            };

            if (num_msgs == 1 and rx_msg.DataSize >= 5) {
                const payload = rx_msg.Data[4..rx_msg.DataSize];
                
                if (payload[0] == 0x7F) {
                    if (payload.len >= 3) {
                        const nrc = payload[2];
                        if (nrc == @intFromEnum(NRC.RESPONSE_PENDING)) {
                            continue;
                        }
                        return error.NegativeResponse;
                    }
                }
                
                const copy_len = @min(payload.len, resp_buffer.len);
                @memcpy(resp_buffer[0..copy_len], payload[0..copy_len]);
                return copy_len;
            }
        }
    }

    pub fn diagnosticSession(self: *UDSClient, session: UDSSession) !void {
        const req = [_]u8{ @intFromEnum(UDSService.DIAGNOSTIC_SESSION_CONTROL), @intFromEnum(session) };
        var resp: [64]u8 = undefined;
        _ = try self.request(&req, &resp);
    }

    pub fn testerPresent(self: *UDSClient) !void {
        const req = [_]u8{ @intFromEnum(UDSService.TESTER_PRESENT), 0x80 }; // 0x80 = suppress positive response
        var resp: [64]u8 = undefined;
        _ = self.request(&req, &resp) catch |err| switch (err) {
            error.Timeout => return, // Timeout is expected if response is suppressed
            else => return err,
        };
    }

    pub fn ecuReset(self: *UDSClient, reset_type: u8) !void {
        const req = [_]u8{ @intFromEnum(UDSService.ECU_RESET), reset_type };
        var resp: [64]u8 = undefined;
        _ = try self.request(&req, &resp);
    }

    pub fn readDataById(self: *UDSClient, did: u16, out_buffer: []u8) !usize {
        const req = [_]u8{ 
            @intFromEnum(UDSService.READ_DATA_BY_ID), 
            @intCast((did >> 8) & 0xFF), 
            @intCast(did & 0xFF) 
        };
        var resp: [1024]u8 = undefined;
        const len = try self.request(&req, &resp);
        if (len < 3) return error.InvalidResponse;
        
        const data_len = len - 3;
        const copy_len = @min(data_len, out_buffer.len);
        @memcpy(out_buffer[0..copy_len], resp[3..3 + copy_len]);
        return copy_len;
    }

    pub fn writeDataById(self: *UDSClient, did: u16, data: []const u8) !void {
        var req: [1024]u8 = undefined;
        req[0] = @intFromEnum(UDSService.WRITE_DATA_BY_ID);
        req[1] = @intCast((did >> 8) & 0xFF);
        req[2] = @intCast(did & 0xFF);
        @memcpy(req[3..3 + data.len], data);
        
        var resp: [64]u8 = undefined;
        _ = try self.request(req[0..3 + data.len], &resp);
    }

    pub fn securityAccessSeed(self: *UDSClient, level: u8, out_seed: []u8) !usize {
        const req = [_]u8{ @intFromEnum(UDSService.SECURITY_ACCESS), level };
        var resp: [64]u8 = undefined;
        const len = try self.request(&req, &resp);
        if (len < 2) return error.InvalidResponse;
        
        const seed_len = len - 2;
        const copy_len = @min(seed_len, out_seed.len);
        @memcpy(out_seed[0..copy_len], resp[2..2 + copy_len]);
        return copy_len;
    }

    pub fn securityAccessKey(self: *UDSClient, level: u8, key: []const u8) !void {
        var req: [64]u8 = undefined;
        req[0] = @intFromEnum(UDSService.SECURITY_ACCESS);
        req[1] = level;
        @memcpy(req[2..2 + key.len], key);
        
        var resp: [64]u8 = undefined;
        _ = try self.request(req[0..2 + key.len], &resp);
    }

    pub fn clearDtc(self: *UDSClient, group: u32) !void {
        const req = [_]u8{
            @intFromEnum(UDSService.CLEAR_DTC),
            @intCast((group >> 16) & 0xFF),
            @intCast((group >> 8) & 0xFF),
            @intCast(group & 0xFF)
        };
        var resp: [64]u8 = undefined;
        _ = try self.request(&req, &resp);
    }

    pub fn readDtc(self: *UDSClient, subfunc: u8, mask: u8, out_buffer: []u8) !usize {
        const req = [_]u8{ @intFromEnum(UDSService.READ_DTC_INFO), subfunc, mask };
        var resp: [1024]u8 = undefined;
        const len = try self.request(&req, &resp);
        const copy_len = @min(len, out_buffer.len);
        @memcpy(out_buffer[0..copy_len], resp[0..copy_len]);
        return copy_len;
    }

    pub fn routineControl(self: *UDSClient, control_type: u8, routine_id: u16, data: []const u8, out_buffer: []u8) !usize {
        var req: [1024]u8 = undefined;
        req[0] = @intFromEnum(UDSService.ROUTINE_CONTROL);
        req[1] = control_type;
        req[2] = @intCast((routine_id >> 8) & 0xFF);
        req[3] = @intCast(routine_id & 0xFF);
        @memcpy(req[4..4 + data.len], data);
        
        var resp: [1024]u8 = undefined;
        const len = try self.request(req[0..4 + data.len], &resp);
        const copy_len = @min(len, out_buffer.len);
        @memcpy(out_buffer[0..copy_len], resp[0..copy_len]);
        return copy_len;
    }

    pub fn ioControl(self: *UDSClient, did: u16, control_param: u8, state: []const u8) !void {
        var req: [64]u8 = undefined;
        req[0] = @intFromEnum(UDSService.IO_CONTROL);
        req[1] = @intCast((did >> 8) & 0xFF);
        req[2] = @intCast(did & 0xFF);
        req[3] = control_param;
        @memcpy(req[4..4 + state.len], state);
        
        var resp: [64]u8 = undefined;
        _ = try self.request(req[0..4 + state.len], &resp);
    }
};
